### PR TITLE
Have speedreader fallback on chrome distill

### PR DESCRIPTION
**Pull request by keur**
If the user activates speedreader and the page is not in the speedreader
whitelist, quietly fallback on Chrome's distiller.

Resolves https://github.com/brave/brave-browser/issues/11994

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/speedreader/BUILD.gn [hunk 1]
```diff
@@ -35,6 +35,7 @@ source_set("speedreader") {
   ]
 
   deps = [
+    "//components/dom_distiller/content/browser",
```
Comment (iefremov) [inline components/speedreader/BUILD.gn:38]: we need it in `browser/BUILD.gn`, not here

**Comment by iefremov** [comment]
Do I understand correctly that we want users to enable Distiller on `brave://flags` to make it tick?

**Comment by iefremov** [comment]
From the user perspective, for now there is no way to undo distillation and get back to the normal page: neither toggling the speedreader button, nor reloading don't bring me back

**Comment by iefremov** [comment]
Also we will have UI discrepancy: pages optimized with our speedreader have normal "lock" icon in the omnibox, while chrome-distilled ones would have "Reader Mode". Is it intended @rebron ?

**Comment by iefremov** [comment]
There is also a trouble with back-forward navigation: with our speedreader hitting "Back" works as expected, while trying to go back on chrome-distilled page brings you back and then immeditealy re-runs the distiller, so you end up on a distilled version once again

File: browser/speedreader/speedreader_service_factory.cc [hunk 2]
```diff
@@ -26,7 +27,9 @@ SpeedreaderService* SpeedreaderServiceFactory::GetForProfile(Profile* profile) {
 SpeedreaderServiceFactory::SpeedreaderServiceFactory()
     : BrowserContextKeyedServiceFactory(
           "SpeedreaderService",
-          BrowserContextDependencyManager::GetInstance()) {}
+          BrowserContextDependencyManager::GetInstance()) {
+  DependsOn(dom_distiller::DomDistillerServiceFactory::GetInstance());
```
Comment (iefremov) [inline browser/speedreader/speedreader_service_factory.cc:31]: not sure if this is really needed

File: browser/speedreader/speedreader_tab_helper.cc [hunk 3]
```diff
@@ -5,21 +5,38 @@
...
 #include "brave/browser/speedreader/speedreader_tab_helper.h"
 
+#include <memory>
+
 #include "brave/browser/brave_browser_process_impl.h"
 #include "brave/browser/speedreader/speedreader_service_factory.h"
 #include "brave/components/speedreader/speedreader_rewriter_service.h"
 #include "brave/components/speedreader/speedreader_service.h"
 #include "brave/components/speedreader/speedreader_test_whitelist.h"
+#include "chrome/browser/dom_distiller/dom_distiller_service_factory.h"
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.cc:15]: not needed

File: browser/ui/views/toolbar/speedreader_button.cc [hunk 4]
```diff
@@ -73,7 +73,7 @@ void SpeedreaderButton::Update(content::WebContents* active_contents) {
     auto* tab_helper =
         speedreader::SpeedreaderTabHelper::FromWebContents(active_contents);
     if (tab_helper) {
-      const bool active = tab_helper->IsActiveForMainFrame();
+      const bool active = tab_helper->IsReaderEnabled();
```
Comment (iefremov) [inline browser/ui/views/toolbar/speedreader_button.cc:76]: mb it should be `IsDistilled`

File: browser/speedreader/speedreader_tab_helper.cc [hunk 5]
```diff
@@ -5,21 +5,38 @@
...
 SpeedreaderTabHelper::~SpeedreaderTabHelper() = default;
 
 SpeedreaderTabHelper::SpeedreaderTabHelper(content::WebContents* web_contents)
     : content::WebContentsObserver(web_contents) {}
 
+bool SpeedreaderTabHelper::IsReaderEnabled() const {
+  if (speedreader_active_)
+    return true;
+  auto url = web_contents()->GetURL();
+  return IsDistilledPage(url);
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.cc:37]: `GetURL()` is deprecated
`return IsDistilledPage(web_contents()->GetLastCommitedURL())`

File: browser/speedreader/speedreader_tab_helper.cc [hunk 6]
```diff
@@ -64,6 +86,21 @@ void SpeedreaderTabHelper::DidRedirectNavigation(
   }
 }
 
+void SpeedreaderTabHelper::OnResult(
+    const dom_distiller::DistillabilityResult& result) {
+  auto url = web_contents()->GetURL();
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.cc:91]: `GetURL()` is deprecated

**Comment by iefremov** [comment]
Going back to a chrome-distilled page results in an infinite throbbler

**Review by iefremov — CHANGES_REQUESTED** [review]
I think we need a bit more detailed product vision for this (cc @rebron)

**Comment by iefremov** [comment]
@keur do you think it's worth finalizing or maybe we could abandon?

**Comment by keur** [comment]
Abandon

**Pull request closed without merging**